journal.py

```diff
@@ -0,0 +1,212 @@
+def edit_view(request):
+    entry = Entry.one(request.matchdict['id'])
+    return {'entry': entry}
+
+
+@view_config(route_name='add', request_method='POST')
+def add_entry(request):
+    title = request.params.get('title')
+    text = request.params.get('text')
+    Entry.write(title=title, text=text)
+    return HTTPFound(request.route_url('home'))
+
+
+@view_config(route_name='modify', request_method='POST')
+def modify_entry(request):
+    eid = request.matchdict['id']
+    title = request.params.get('title')
+    text = request.params.get('text')
+    Entry.modify(eid=eid, title=title, text=text)
+    return HTTPFound(request.route_url('home'))
+
+
+@view_config(route_name='login', renderer='templates/login.jinja2')
+def login(request):
+    """authenticate a user by username/password"""
+    username = request.params.get('username', '')
+    error = ''
+    if request.method == 'POST':
+        error = "Login Failed"
+        authenticated = False
+        try:
+            authenticated = do_login(request)
+        except ValueError as e:
+            error = str(e)
+
+        if authenticated:
+            headers = remember(request, username)
+            return HTTPFound(request.route_url('home'), headers=headers)
+
+    return {'error': error, 'username': username}
+
+
+@view_config(route_name='logout')
+def logout(request):
+    headers = forget(request)
+    return HTTPFound(request.route_url('home'), headers=headers)
+
+
+@view_config(context=DBAPIError)
+def db_exception(context, request):
+    from pyramid.response import Response
+    response = Response(context.message)
+    response.status_int = 500
+    return response
+
+
+def main():
+    """Create a configured wsgi app"""
+    settings = {}
+    debug = os.environ.get('DEBUG', True)
+    settings['reload_all'] = debug
+    settings['debug_all'] = debug
+    settings['auth.username'] = os.environ.get('AUTH_USERNAME', 'admin')
+    manager = BCRYPTPasswordManager()
+    settings['auth.password'] = os.environ.get(
+        'AUTH_PASSWORD', manager.encode('secret')
+    )
+    if not os.environ.get('TESTING', False):
+        # only bind the session if we are not testing
+        engine = sa.create_engine(DATABASE_URL)
+        DBSession.configure(bind=engine)
+    # add a secret value for auth tkt signing
+    auth_secret = os.environ.get('JOURNAL_AUTH_SECRET', 'itsaseekrit')
+    # configuration setup
+    config = Configurator(
+        settings=settings,
+        authentication_policy=AuthTktAuthenticationPolicy(
+            secret=auth_secret,
+            hashalg='sha512'
+        ),
+        authorization_policy=ACLAuthorizationPolicy(),
+    )
+    config.include('pyramid_tm')
+    config.include('pyramid_jinja2')
+    config.add_route('home', '/')
+    config.add_route('add', '/add')
+    config.add_route('login', '/login')
+    config.add_route('logout', '/logout')
+    config.add_route('detail', '/detail/{id}')
```

Comment:
good use of a URL path and placeholders to pass the id of the desired entry around.